rust/template/differential_datalog/program.rs

```diff
@@ -2594,29 +2769,37 @@ impl RunningProgram {
         Ok(())
     }
 
-    /* Reverse all changes recorded in delta sets to rollback the transaction. */
-    fn delta_undo(&mut self) -> Response<()> {
-        let mut updates = vec![];
-        for (relid, rel) in &self.relations {
-            // first delete, then insert to avoid duplicate key
-            // errors in `apply_updates()`
-            for (k, w) in rel.delta() {
-                if *w {
+    fn delta_undo_updates(relid: RelId, ds: &DeltaSet, updates: &mut Vec<Update<DDValue>>) {
+        // first delete, then insert to avoid duplicate key
+        // errors in `apply_updates()`
+        for (k, w) in ds {
+            if *w >= 0 {
+                for _ in 0..*w {
```

Comment:
this looks inefficient; weights could potentially be very large

Reply:
I know what you mean but in this particular place it's not too bad, as there will be at most one compensation command for each input command from the user.